analyzers/src/SonarAnalyzer.Common/AnalysisContext/DiagnosticAnalyzerContextHelper.cs

```diff
@@ -0,0 +1,143 @@
+                                                              Action<SyntaxTreeAnalysisContext> action) =>
+        context.RegisterCompilationStartAction(csac =>
+            csac.RegisterSyntaxTreeAction(c =>
+                {
+                    if (SonarAnalysisContext.ShouldAnalyze(context.Context.TryGetValue, context.Context.TryGetValue, generatedCodeRecognizer, c.GetSyntaxTree(), csac.Compilation, c.Options))
+                    {
+                        action(c);
+                    }
+                }));
+
+    public static void RegisterCodeBlockStartActionInNonGenerated<TLanguageKindEnum>(this SonarAnalysisContext context,
+                                                                                     GeneratedCodeRecognizer generatedCodeRecognizer,
+                                                                                     Action<CodeBlockStartAnalysisContext<TLanguageKindEnum>> action) where TLanguageKindEnum : struct =>
+        context.RegisterCodeBlockStartAction<TLanguageKindEnum>(c =>
+            {
+                if (SonarAnalysisContext.ShouldAnalyze(context.TryGetValue, context.TryGetValue, generatedCodeRecognizer, c.GetSyntaxTree(), c.SemanticModel.Compilation, c.Options))
+                {
+                    action(c);
+                }
+            });
+
+    public static void ReportDiagnosticIfNonGenerated(this CompilationAnalysisContext context, GeneratedCodeRecognizer generatedCodeRecognizer, Diagnostic diagnostic)
+    {
+        if (SonarAnalysisContext.ShouldAnalyze(context.TryGetValue, context.TryGetValue, generatedCodeRecognizer, diagnostic.Location.SourceTree, context.Compilation, context.Options))
+        {
+            context.ReportIssue(diagnostic);
+        }
+    }
+
+    public static void ReportDiagnosticIfNonGenerated(this SymbolAnalysisContext context, GeneratedCodeRecognizer generatedCodeRecognizer, Diagnostic diagnostic)
+    {
+        if (SonarAnalysisContext.ShouldAnalyze(AnalyzeGeneratedNoCache, ProjectConfigReaderNoCache, generatedCodeRecognizer, diagnostic.Location.SourceTree, context.Compilation, context.Options))
+        {
+            context.ReportIssue(diagnostic);
+        }
+
+        bool AnalyzeGeneratedNoCache(SourceText text, SourceTextValueProvider<bool> valueProvider, out bool value)
+        {
+            value = PropertiesHelper.ReadAnalyzeGeneratedCodeProperty(PropertiesHelper.GetSettings(context.Options), context.Compilation.Language);
+            return true;
+        }
+
+        bool ProjectConfigReaderNoCache(SourceText text, SourceTextValueProvider<ProjectConfigReader> valueProvider, out ProjectConfigReader value)
+        {
+            value = SonarAnalysisContext.ProjectConfiguration(CreateProjectConfigReader, context.Options);
+            return true;
+        }
+
+            bool CreateProjectConfigReader(SourceText text, SourceTextValueProvider<ProjectConfigReader> valueProvider, out ProjectConfigReader value)
```

Comment:
The indentation is off here.

Reply:
This (ugly) part will be deleted anyway, I won't fix it on this PR.